operator/pkg/lbipam/lbipam.go

```diff
@@ -0,0 +1,1728 @@
+		poolResource:      params.PoolResource,
+		svcResource:       params.SvcResource,
+		initDoneCallbacks: params.NotifyReadyCallbacks,
+		poolClient:        params.ClientSet.CiliumV2alpha1().CiliumLoadBalancerIPPools(),
+		svcClient:         params.ClientSet.CoreV1(),
+		poolStore:         NewPoolStore(),
+		rangesStore:       newRangesStore(),
+		serviceStore:      NewServiceStore(),
+		lbClasses:         lbClasses,
+		ipv4Enabled:       option.Config.IPv4Enabled(),
+		ipv6Enabled:       option.Config.IPv6Enabled(),
+	}
+
+	// Make a new Ctx which we will use to manage the lifetime of LBIPAM.Run.
+	ipamCtx, cancel := context.WithCancel(context.Background())
+	var wg sync.WaitGroup
+
+	params.LC.Append(hive.Hook{
+		OnStart: func(ctx hive.HookContext) error {
+			params.Logger.Info("Starting LB IPAM")
+
+			wg.Add(1)
+
+			// Start go routine, we can't block in OnStart
+			go func() {
+				defer wg.Done()
+				lbIPAM.Run(ipamCtx)
+			}()
+
+			return nil
+		},
+		OnStop: func(ctx hive.HookContext) error {
+			// Close the ipamCtx
+			cancel()
+
+			// Convert blocking Wait into a channel so we can select on the context closing or the Wait.
+			wgDone := make(chan struct{})
+			go func() {
+				wg.Wait()
+				close(wgDone)
+			}()
+
+			select {
+			case <-wgDone:
+			case <-ctx.Done():
+			}
+
+			return nil
+		},
+	})
+
+	return lbIPAM
+}
+
+// LBIPAM is the loadbalancer IP address manager, watcher/controller which allocates and assigns IP addresses
+// to LoadBalancer services from the configured set of LoadBalancerIPPools in the cluster.
+type LBIPAM struct {
+	logger logrus.FieldLogger
+
+	lbClasses   []string
+	ipv4Enabled bool
+	ipv6Enabled bool
+
+	poolClient cilium_client_v2alpha1.CiliumLoadBalancerIPPoolInterface
+	svcClient  client_typed_v1.ServicesGetter
+
+	poolResource resource.Resource[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool]
+	svcResource  resource.Resource[*core_v1.Service]
+
+	poolStore    poolStore
+	rangesStore  rangesStore
+	serviceStore serviceStore
+
+	ctx context.Context
+
+	// Only used during testing.
+	initDoneCallbacks []LBIPAMInitDone
+}
+
+type ipPoolEvent = resource.Event[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool]
+type svcEvent = resource.Event[*core_v1.Service]
+
+func (ipam *LBIPAM) Run(ctx context.Context) {
+	ipam.ctx = ctx
+
+	errChan := make(chan error, 2)
+
+	poolChan := stream.ToChannel[ipPoolEvent](ctx, errChan, ipam.poolResource)
+	svcChan := stream.ToChannel[svcEvent](ctx, errChan, ipam.svcResource)
+
+	err := ipam.init(ctx, poolChan, svcChan)
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error while initializing LB-IPAM")
+		// Stop, don't go into main control loop
+		return
+	}
+
+	for _, cb := range ipam.initDoneCallbacks {
+		if cb != nil {
+			cb()
+		}
+	}
+
+	// When the ctx expires, both streams will close, keep processing until both channels are closed to avoid blocking
+	// the resources in case they are re-used after we shutdown.
+	for poolChan != nil && svcChan != nil {
```

Comment:
This will stop processing as soon as one of the channels is closed; did you mean `poolChan != nil || svcChan != nil` ?

Reply:
No, the resources expect both channels to be drained on shutdown, this ensures we keep reading until we get a close signal from both channels. I will reword the comment to make this clearer

Reply:
Right; I think you have the logic backwards. It should be a `||` and not an `&&`.

Reply:
I think @squeed is right. We need `poolChan != nil || svcChan != nil`, e.g. keep going until both are non-nil.

Reply:
Ah, I see it now. Sorry about that one